{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene-Prediction-in-Metagenomics-Fragments\n",
    "Written by: Vignesh Ravindranath, Shawn Huang\n",
    "\n",
    "This project was based on the 2008 paper ['Gene prediction in metagenomic fragments: A large scale machine learning approach'](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-217) by Hoff et at. The purpose of the project is to gain a `deeper understanding of machine learning approaches behind the Orphelia gene prediction algorithm`. \n",
    "    \n",
    "    The project consists of several steps:\n",
    "        1) Exploring SeqRecord Data and Structure (Genbank, Fasta)\n",
    "        2) Feature Preprocessing\n",
    "        3) Linear Discriminant training on high-dimentional features\n",
    "        4) Binary classification neural network for coding/noncoding gene prediction\n",
    "    \n",
    "In the second step, high-dimensional features such as `monocodon (tricodon) usage`, `dicodon (hexcodon) usage`, and `translation initiation sites (TIS)` must be extracted from prokaryotic genomes. Monocodon and dicodon usage refer to the frequence of codons (3 basepairs) and dicodons (6 basepairs) in coding and noncoding regions. Coding proteins often have a TIS upstream of the start codon. These TIS patterns are extracted by comparing up and downstream regions for positive TIS candidates (true start codons in coding sequences) to negative TIS candidates (start codons in noncoding sequences).  \n",
    "    \n",
    "In the third step, linear discriminants are derived to reduce the dimensionality of the extracted features. The individual features (excluding GC content) are taken as multivariate linear regression problems and the Normal Equation is utilized to compute the weights (coefficient) matrix for each feature. \n",
    "    \n",
    "In the last step, a neural network is trained on the reduced features mentioned above, in addition to the `length of the sequence` and the `GC content`.\n",
    "\n",
    "    Summary of features:\n",
    "    x1 - tricodon       - (n,64) table --reduced to a weights matrix of (64,1)\n",
    "    x2 - hexcodon       - (n,4096) table --reduced to a weights matrix of (4096,1)\n",
    "    x3 - positive TIS   - (n,58,64 == n,3712) table --reduced to a weights matrix of (3712,1)\n",
    "    x4 - negative TIS   - (m,58,64 == m,3712) table --reduced to a weights matrix of (3712,1)\n",
    "    x5 - complete seq   - 1 if fragment contains a complete gene, else 0\n",
    "    x6 - incomplete seq - 0 if fragment contains a complete gene, else 1\n",
    "    x7 - GC content     - (n,1) column"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test prokaryotic genome: [A fulgidus](https://www.ncbi.nlm.nih.gov/nuccore/NC_000917)\n",
    "\n",
    "Biopython documentation: \n",
    "[Biopython Tutorial](http://biopython.org/DIST/docs/tutorial/Tutorial.html), \n",
    "[SeqFeature](https://biopython.org/docs/dev/api/Bio.SeqFeature.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tensorflow.keras.utils import to_categorical\n",
    "\n",
    "import re\n",
    "import itertools\n",
    "import textwrap\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import Entrez\n",
    "Entrez.email = \"vignesh.ravindranath@gmail.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched NC_000917 GenBank records\n",
      "Fetched NC_000917 FASTA sequence\n"
     ]
    }
   ],
   "source": [
    "prokaryote_id = 'NC_000917' # A. fulgidus\n",
    "\n",
    "# get seq_records (features) from GenBank file\n",
    "handle = Entrez.efetch(db=\"sequences\", id=prokaryote_id, rettype=\"gbwithparts\", retmode=\"text\")\n",
    "seq_record = SeqIO.read(handle, \"gb\")\n",
    "handle.close()\n",
    "print(f'Fetched {prokaryote_id} GenBank records')\n",
    "\n",
    "# get full sequence from FASTA file\n",
    "handle = Entrez.efetch(db=\"sequences\", id=prokaryote_id, rettype=\"fasta\", retmode=\"text\")\n",
    "sequence = SeqIO.read(handle, \"fasta\")\n",
    "handle.close()\n",
    "print(f'Fetched {prokaryote_id} FASTA sequence')\n",
    "\n",
    "dna = [sequence, sequence.reverse_complement()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exploring SeqRecord Data and Structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2473\n"
     ]
    }
   ],
   "source": [
    "# find coding sequences (cds), or genes, for A. fulgidus\n",
    "cds = []\n",
    "\n",
    "for feature in seq_record.features:\n",
    "    if feature.type == 'CDS':\n",
    "        cds.append(feature)\n",
    "        \n",
    "print(len(cds))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "type: CDS\n",
      "location: [4289:4784](+)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: db_xref, Value: ['GeneID:1483212']\n",
      "    Key: inference, Value: ['COORDINATES: protein motif:HMM:NF012415.1']\n",
      "    Key: locus_tag, Value: ['AF_RS00005']\n",
      "    Key: note, Value: ['Derived by automated computational analysis using gene prediction method: Protein Homology.']\n",
      "    Key: old_locus_tag, Value: ['AF0003', 'AF_0003']\n",
      "    Key: product, Value: ['CAP domain-containing protein']\n",
      "    Key: protein_id, Value: ['WP_010877517.1']\n",
      "    Key: transl_table, Value: ['11']\n",
      "    Key: translation, Value: ['MKETIQLAIGVMLLAMLGCYIYITEFYHYESTEESSKAAIEYLNQLRAQNGLPPVKWNKTLYEFALERLEDMHERGYYSHYDPVTHETLIYRYVEGYVGECILNGVRGTNLLSNGLQSLFGYEEEAIDIWSKSTMHKLILTDKRFTDAAVACKYDMCVLIMTGG']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(cds[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We only need relevant information from the SeqRecord such as:\n",
    "\n",
    "    type (coding sequence)\n",
    "    location (start location, stop location, and strand)\n",
    "    locus_tag (id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_000917.1\n"
     ]
    }
   ],
   "source": [
    "print(seq_record.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4289\n",
      "4784\n",
      "1\n",
      "AF_RS00005\n",
      "MKETIQLAIGVMLLAMLGCYIYITEFYHYESTEESSKAAIEYLNQLRAQNGLPPVKWNKTLYEFALERLEDMHERGYYSHYDPVTHETLIYRYVEGYVGECILNGVRGTNLLSNGLQSLFGYEEEAIDIWSKSTMHKLILTDKRFTDAAVACKYDMCVLIMTGG\n",
      "---check if using correct dna strand\n",
      "MKETIQLAIGVMLLAMLGCYIYITEFYHYESTEESSKAAIEYLNQLRAQNGLPPVKWNKTLYEFALERLEDMHERGYYSHYDPVTHETLIYRYVEGYVGECILNGVRGTNLLSNGLQSLFGYEEEAIDIWSKSTMHKLILTDKRFTDAAVACKYDMCVLIMTGG*\n"
     ]
    }
   ],
   "source": [
    "# Coding sequence on coding strand (+1)\n",
    "\n",
    "print(cds[0].location.start.position)\n",
    "print(cds[0].location.end.position)\n",
    "print(cds[0].strand)\n",
    "print(cds[0].qualifiers['locus_tag'][0])\n",
    "print(cds[0].qualifiers['translation'][0])\n",
    "\n",
    "start = cds[0].location.start.position\n",
    "end = cds[0].location.end.position\n",
    "\n",
    "print('---check if using correct dna strand')\n",
    "seq0 = dna[0][start:end].seq\n",
    "pro = seq0.translate()\n",
    "print(pro)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44802\n",
      "45561\n",
      "-1\n",
      "AF_RS00195\n",
      "MSGYLRKRDVIITMTQFEFKQRYRGTALGLIWSLLAPFLLALVLFLVFRNMFSWVENFAAYVLVGVFVFRFFQVATSVGMHTIVGKSHLVTKTNIDRELLPLATTLSYGMSSFLEILVIVPIVHVLGGNVGFTILFLPVIHFVYIIFIFGLNLFLSSLMVYFRDLNQIWEVITNVIFFASPIVYPLTMIPESYREMYMLNPIACIIEIYRGILMENQILLEKFIYFLLISLALTFAGQLFFRRMQKRFGEVL\n",
      "---check if using correct dna strand\n",
      "MSGYLRKRDVIITMTQFEFKQRYRGTALGLIWSLLAPFLLALVLFLVFRNMFSWVENFAAYVLVGVFVFRFFQVATSVGMHTIVGKSHLVTKTNIDRELLPLATTLSYGMSSFLEILVIVPIVHVLGGNVGFTILFLPVIHFVYIIFIFGLNLFLSSLMVYFRDLNQIWEVITNVIFFASPIVYPLTMIPESYREMYMLNPIACIIEIYRGILMENQILLEKFIYFLLISLALTFAGQLFFRRMQKRFGEVL*\n"
     ]
    }
   ],
   "source": [
    "# Coding sequence on complementary strand (-1)\n",
    "\n",
    "print(cds[39].location.start.position)\n",
    "print(cds[39].location.end.position)\n",
    "print(cds[39].strand)\n",
    "print(cds[39].qualifiers['locus_tag'][0])\n",
    "print(cds[39].qualifiers['translation'][0])\n",
    "\n",
    "start = cds[39].location.start.position\n",
    "end = cds[39].location.end.position\n",
    "\n",
    "print('---check if using correct dna strand')\n",
    "seq39 = dna[1][::-1][start:end][::-1].seq\n",
    "pro = seq39.translate()\n",
    "print(pro)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find Coding Sequences (CDS) or Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2473\n"
     ]
    }
   ],
   "source": [
    "# Get positive ORFs (coding sequences (CDS) or genes)\n",
    "l_min = sys.maxsize\n",
    "cds = []\n",
    "for feature in seq_record.features:\n",
    "    if feature.type == 'CDS':\n",
    "        tag    = feature.qualifiers['locus_tag'][0]\n",
    "        start  = feature.location.start.position\n",
    "        end    = feature.location.end.position\n",
    "        strand = feature.strand\n",
    "        seq    = dna[0][start:end] if strand == 1 else dna[1][::-1][start:end][::-1]\n",
    "\n",
    "        # update min\n",
    "        if len(seq) < l_min:\n",
    "            l_min = len(seq)\n",
    "\n",
    "        f = [SeqFeature(FeatureLocation(start,end,strand), type=\"CDS\")]\n",
    "        r = SeqRecord(seq.seq, name=tag, id=prokaryote_id, features=f)\n",
    "        cds.append(r)\n",
    "        \n",
    "print(len(cds))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: NC_000917\n",
      "Name: AF_RS00005\n",
      "Description: <unknown description>\n",
      "Number of features: 1\n",
      "Seq('ATGAAAGAGACGATTCAGCTTGCTATAGGGGTGATGCTGTTGGCCATGCTCGGT...TGA')\n"
     ]
    }
   ],
   "source": [
    "print(cds[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "type: CDS\n",
      "location: [4289:4784](+)\n",
      "qualifiers:\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(cds[0].features[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find Longest Open Reading Frame (ORF)\n",
    "    \n",
    "An open reading frame (ORF) is part of a reading frame that has the ability to translate; it will begin with a start codon (ATG) and end with an in-frame stop codon (TAG, TGA, TAA). Prokaryotes may have alternate start codons, such as CTG, GTG, and TTG. The code below searches a sequences for all ORFs and returns the longest sequence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prokaryotes may have alternate start codons\n",
    "START_CODONS = ['ATG','CTG','GTG','TTG']\n",
    "STOP_CODONS = ['TAG','TGA','TAA']\n",
    "    \n",
    "def longest_orf(seq):\n",
    "    all_starts = []\n",
    "    for s in START_CODONS:\n",
    "        # Find positions of all start codons in sequence\n",
    "        matches = re.finditer(s, str(seq))\n",
    "        matches_positions = [match.start() for match in matches]\n",
    "        all_starts.extend(matches_positions)\n",
    "    all_starts = sorted(all_starts)\n",
    "\n",
    "    all_stops = []                                      \n",
    "    for e in STOP_CODONS:\n",
    "        # Find positions of all stop codons in sequence\n",
    "        matches = re.finditer(e, str(seq))\n",
    "        matches_positions = [match.start() for match in matches]\n",
    "        all_stops.extend(matches_positions)\n",
    "    all_stops = sorted(all_stops)\n",
    "\n",
    "    # find largest ORF\n",
    "    for s in all_starts:\n",
    "        for e in all_stops[::-1]:\n",
    "            if (e >= s) and ((e-s)%3 == 0):\n",
    "                return [s,e+3]\n",
    "            \n",
    "    return []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGAAAGAGACGATTCAGCTTGCTATAGGGGTGATGCTGTTGGCCATGCTCGGTTGCTACATCTACATTACTGAGTTCTATCACTACGAATCAACTGAAGAGAGCTCAAAAGCTGCGATTGAATATTTGAATCAGCTTAGAGCTCAGAATGGCCTGCCTCCTGTAAAGTGGAATAAAACTCTTTACGAGTTTGCGCTGGAAAGGCTTGAAGATATGCATGAAAGAGGCTATTACAGTCATTATGACCCTGTTACGCATGAGACGCTGATTTACAGATATGTTGAGGGGTATGTTGGAGAGTGTATCTTGAATGGTGTAAGAGGTACAAATCTTCTCTCCAATGGGCTTCAGTCATTATTTGGCTATGAAGAAGAAGCTATAGATATCTGGTCTAAAAGCACCATGCACAAACTCATTCTCACTGATAAACGCTTCACAGATGCCGCTGTAGCCTGCAAGTACGACATGTGTGTTTTGATTATGACGGGTGGTTGA\n",
      "\n",
      "ATGAAAGAGACGATTCAGCTTGCTATAGGGGTGATGCTGTTGGCCATGCTCGGTTGCTACATCTACATTACTGAGTTCTATCACTACGAATCAACTGAAGAGAGCTCAAAAGCTGCGATTGAATATTTGAATCAGCTTAGAGCTCAGAATGGCCTGCCTCCTGTAAAGTGGAATAAAACTCTTTACGAGTTTGCGCTGGAAAGGCTTGAAGATATGCATGAAAGAGGCTATTACAGTCATTATGACCCTGTTACGCATGAGACGCTGATTTACAGATATGTTGAGGGGTATGTTGGAGAGTGTATCTTGAATGGTGTAAGAGGTACAAATCTTCTCTCCAATGGGCTTCAGTCATTATTTGGCTATGAAGAAGAAGCTATAGATATCTGGTCTAAAAGCACCATGCACAAACTCATTCTCACTGATAAACGCTTCACAGATGCCGCTGTAGCCTGCAAGTACGACATGTGTGTTTTGATTATGACGGGTGGTTGA\n"
     ]
    }
   ],
   "source": [
    "start = cds[0].features[0].location.start.position\n",
    "end = cds[0].features[0].location.end.position\n",
    "test = dna[0][start:end].seq\n",
    "\n",
    "orf = longest_orf(test)\n",
    "\n",
    "print(test)\n",
    "print()\n",
    "print(dna[0][orf[0]+start:orf[1]+start].seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGTTTATGGGGTAG\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "#        AAA ATG TTT ATG GGG TAG \n",
    "test2 = Seq('AAAATGTTTATGGGGTAG')\n",
    "\n",
    "# result:    ATG TTT ATG GGG TAG\n",
    "orf = longest_orf(test2)\n",
    "print(test2[orf[0]:orf[1]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find Noncoding Sequences (NCS) in Interregions\n",
    "Next, we need to find the `negative ORFs` or the ORFs found in interregions (non coding regions). To do this, we will borrow code from [Iddo Friedberg and Ian MC Fleming](https://biopython.org/wiki/Intergenic_regions),  to isolate the interregions for a genome. Then, we will use a function below (`orf_finder`) based on `longest_orf` to search for all in-frame ORF in that sequence, if any."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find interregions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copyright(C) 2009 Iddo Friedberg & Ian MC Fleming\n",
    "# Released under Biopython license. http://www.biopython.org/DIST/LICENSE\n",
    "# Do not remove this comment\n",
    "def get_interregions(coding_records):\n",
    "    cds_list_plus = []\n",
    "    cds_list_minus = []\n",
    "    intergenic_records = []\n",
    "    \n",
    "    initials = coding_records[0].name[0:2]\n",
    "    prokaryote_id = seq_record.id\n",
    "\n",
    "    # Loop over the genome file, get the CDS features on each of the strands\n",
    "    for record in coding_records:\n",
    "        feature = record.features[0]\n",
    "        mystart = feature.location.start.position\n",
    "        myend = feature.location.end.position\n",
    "        if feature.strand == -1:\n",
    "            cds_list_minus.append((mystart,myend,-1))\n",
    "        elif feature.strand == 1:\n",
    "            cds_list_plus.append((mystart,myend,1))\n",
    "        else:\n",
    "            sys.stderr.write(\"No strand indicated %d-%d. Assuming +\\n\" %(mystart, myend))\n",
    "            cds_list_plus.append((mystart,myend,1))\n",
    "    for i,pospair in enumerate(cds_list_plus[1:]):\n",
    "        # Compare current start position to previous end position\n",
    "        last_end = cds_list_plus[i][1]\n",
    "        this_start = pospair[0]\n",
    "        if this_start - last_end >= 1:\n",
    "            intergene_seq = seq_record.seq[last_end:this_start]\n",
    "            strand_string = +1\n",
    "            name = initials + '_NC' + str(i).zfill(5)\n",
    "            feature = [SeqFeature(FeatureLocation(last_end+1,this_start,strand_string), type='interregion')]\n",
    "            intergenic_records.append(SeqRecord(intergene_seq, name=prokaryote_id, id=name, features=feature))\n",
    "    buffer = i\n",
    "    for i,pospair in enumerate(cds_list_minus[1:]):\n",
    "        last_end = cds_list_minus[i][1]\n",
    "        this_start = pospair[0]\n",
    "        if this_start - last_end >= 1:\n",
    "            intergene_seq = seq_record.seq[last_end:this_start]\n",
    "            strand_string = -1\n",
    "            name = initials + '_NC' + str(i+buffer).zfill(5)\n",
    "            feature = [SeqFeature(FeatureLocation(last_end+1,this_start,strand_string), type='interregion')]\n",
    "            intergenic_records.append(SeqRecord(intergene_seq, name=prokaryote_id, id=name, features=feature))\n",
    "    return intergenic_records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1797\n"
     ]
    }
   ],
   "source": [
    "interregions = get_interregions(cds)\n",
    "print(len(interregions))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: AF_NC00000\n",
      "Name: NC_000917.1\n",
      "Description: <unknown description>\n",
      "Number of features: 1\n",
      "Seq('GGTGTTCAGAAAAAAGAGGAAAAGTAAAAAACAATTGCTGCGATAATCGTTTTT...CCA')\n",
      "type: interregion\n",
      "location: [4785:5050](+)\n",
      "qualifiers:\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(interregions[0])\n",
    "print(interregions[0].features[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find longest ORF in interregions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AF_NC00000\n",
      "GTGTTCAGAAAAAAGAGGAAAAGTAAAAAACAATTGCTGCGATAATCGTTTTTGCCGGCGTTTTTTGAAATCAGACTAAAGTGGGATTACGTAAATTTATTTGTGGAAAGTGTGACAAGGAAACAAAACTCAGACCGACGAGAGTTTAAAACCGCCTAAAGCTCTTCTCCAGCAAATCCACCACCAAATTTATTCATTAAGCTTTAACCGTTTAAAAATCCCCAAACGCCCAAACCTTAA\n"
     ]
    }
   ],
   "source": [
    "test3 = interregions[0].seq\n",
    "orf = longest_orf(test3)\n",
    "print(interregions[0].id)\n",
    "print(test3[orf[0]:orf[1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AF_NC00002\n",
      "TTGTTATGAACTGATTCATAA\n"
     ]
    }
   ],
   "source": [
    "test4 = interregions[2].seq\n",
    "orf = longest_orf(test4)\n",
    "print(interregions[2].id)\n",
    "print(test4[orf[0]:orf[1]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find in-frame ORFs - orf-finder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "START_CODONS = ['ATG','CTG','GTG','TTG']\n",
    "STOP_CODONS = ['TAG','TGA','TAA']\n",
    "\n",
    "def codon_pos(seq, codon_list):\n",
    "    pos = []\n",
    "    for codon in codon_list:\n",
    "        matches = re.finditer(codon, str(seq))\n",
    "        matches_positions = [match.start() for match in matches]\n",
    "        pos.extend(matches_positions)\n",
    "    return sorted(pos)\n",
    "\n",
    "'''\n",
    "Input: Bio.Seq.Seq\n",
    "Output: returns start and end position of longest orf in a sequence\n",
    "'''\n",
    "def longest_orf(seq):\n",
    "    all_starts = codon_pos(seq, START_CODONS)\n",
    "    all_stops  = codon_pos(seq, STOP_CODONS)\n",
    "           \n",
    "    for s in all_starts:\n",
    "        for e in all_stops[::-1]:\n",
    "            if (e >= s) and ((e-s)%3 == 0):\n",
    "                return [s,e+3]\n",
    "    return []\n",
    "\n",
    "'''\n",
    "Input: Bio.Seq.Seq\n",
    "Output: returns all in-frame start codons within an ORF sequence\n",
    "'''\n",
    "def orf_finder(orf_seq):\n",
    "    all_starts = codon_pos(orf_seq, START_CODONS)\n",
    "\n",
    "    # find all ORF\n",
    "    orfs = []           \n",
    "    for s in all_starts:\n",
    "        if s%3 == 0:\n",
    "            orfs.append([s,len(orf_seq)])\n",
    "    return orfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0, 495], [30, 495], [33, 495], [36, 495], [39, 495]]\n",
      "ATG\n",
      "GTG\n",
      "ATG\n",
      "CTG\n",
      "TTG\n"
     ]
    }
   ],
   "source": [
    "test5 = cds[0].seq\n",
    "\n",
    "orfs = orf_finder(test5)\n",
    "print(orfs[0:5])\n",
    "\n",
    "for orf in orfs[0:5]:\n",
    "    print(test5[orf[0]:orf[0]+3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0, 240], [33, 240], [36, 240], [51, 240], [102, 240]]\n",
      "GTG\n",
      "TTG\n",
      "CTG\n",
      "TTG\n",
      "GTG\n"
     ]
    }
   ],
   "source": [
    "test6 = ncs[0].seq\n",
    "\n",
    "orfs = orf_finder(test6)\n",
    "print(orfs)\n",
    "\n",
    "for orf in orfs:\n",
    "    print(test6[orf[0]:orf[0]+3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature Preprocessing ORFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copyright(C) 2009 Iddo Friedberg & Ian MC Fleming\n",
    "# Released under Biopython license. http://www.biopython.org/DIST/LICENSE\n",
    "# Do not remove this comment\n",
    "def get_interregions(coding_records):\n",
    "    cds_list_plus = []\n",
    "    cds_list_minus = []\n",
    "    intergenic_records = []\n",
    "    \n",
    "    initials = coding_records[0].name[0:2]\n",
    "    prokaryote_id = coding_records[0].id\n",
    "\n",
    "    # Loop over the genome file, get the CDS features on each of the strands\n",
    "    for record in coding_records:\n",
    "        feature = record.features[0]\n",
    "        mystart = feature.location.start.position\n",
    "        myend = feature.location.end.position\n",
    "        if feature.strand == -1:\n",
    "            cds_list_minus.append((mystart,myend,-1))\n",
    "        elif feature.strand == 1:\n",
    "            cds_list_plus.append((mystart,myend,1))\n",
    "        else:\n",
    "            sys.stderr.write(\"No strand indicated %d-%d. Assuming +\\n\" %(mystart, myend))\n",
    "            cds_list_plus.append((mystart,myend,1))\n",
    "    for i,pospair in enumerate(cds_list_plus[1:]):\n",
    "        # Compare current start position to previous end position\n",
    "        last_end = cds_list_plus[i][1]\n",
    "        this_start = pospair[0]\n",
    "        if this_start - last_end >= 1:\n",
    "            intergene_seq = seq_record.seq[last_end:this_start]\n",
    "            strand_string = +1\n",
    "            tag = initials + '_NC' + str(i).zfill(5)\n",
    "            feature = [SeqFeature(FeatureLocation(last_end+1,this_start,strand_string), type='interregion')]\n",
    "            intergenic_records.append(SeqRecord(intergene_seq, name=tag, id=prokaryote_id, features=feature))\n",
    "    buffer = i\n",
    "    for i,pospair in enumerate(cds_list_minus[1:]):\n",
    "        last_end = cds_list_minus[i][1]\n",
    "        this_start = pospair[0]\n",
    "        if this_start - last_end >= 1:\n",
    "            intergene_seq = seq_record.seq[last_end:this_start]\n",
    "            strand_string = -1\n",
    "            tag = initials + '_NC' + str(i+buffer).zfill(5)\n",
    "            feature = [SeqFeature(FeatureLocation(last_end+1,this_start,strand_string), type='interregion')]\n",
    "            intergenic_records.append(SeqRecord(intergene_seq, name=tag, id=prokaryote_id, features=feature))\n",
    "    return intergenic_records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "START_CODONS = ['ATG','CTG','GTG','TTG']\n",
    "STOP_CODONS = ['TAG','TGA','TAA']\n",
    "\n",
    "def codon_pos(seq, codon_list):\n",
    "    pos = []\n",
    "    for codon in codon_list:\n",
    "        matches = re.finditer(codon, str(seq))\n",
    "        matches_positions = [match.start() for match in matches]\n",
    "        pos.extend(matches_positions)\n",
    "    return sorted(pos)\n",
    "\n",
    "'''\n",
    "Input: Bio.Seq.Seq\n",
    "Output: returns start and end position of longest orf in a sequence\n",
    "'''\n",
    "def longest_orf(seq):\n",
    "    all_starts = codon_pos(seq, START_CODONS)\n",
    "    all_stops  = codon_pos(seq, STOP_CODONS)\n",
    "\n",
    "    orfs = []\n",
    "    found = False;            \n",
    "    for s in all_starts:\n",
    "        for e in all_stops[::-1]:\n",
    "            if (e >= s) and ((e-s)%3 == 0):\n",
    "                found = True; orfs = [s,e+3];\n",
    "                break\n",
    "        if found: break\n",
    "            \n",
    "    return orfs\n",
    "\n",
    "'''\n",
    "Input: Bio.Seq.Seq\n",
    "Output: returns all in-frame start codons within an ORF sequence\n",
    "'''\n",
    "def orf_finder(orf_seq):\n",
    "    all_starts = codon_pos(orf_seq, START_CODONS)\n",
    "\n",
    "    # find all ORF\n",
    "    orfs = []\n",
    "    e = len(orf_seq)\n",
    "    for s in all_starts:\n",
    "        if (e >= s) and (s%3 == 0):\n",
    "            orfs.append([s,e])\n",
    "    return orfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_genome(seq_record, dna):\n",
    "    # Get positive ORFs (coding sequences (CDS) or genes)\n",
    "    l_min = 50\n",
    "    cds = []\n",
    "    genes = []\n",
    "    for feature in seq_record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            tag    = feature.qualifiers['locus_tag'][0]\n",
    "            start  = feature.location.start.position\n",
    "            end    = feature.location.end.position\n",
    "            strand = feature.strand\n",
    "            seq    = dna[0][start:end] if strand == 1 else dna[1][::-1][start:end][::-1]\n",
    "            \n",
    "            orfs = orf_finder(seq.seq)\n",
    "            \n",
    "            for i,orf in enumerate(orfs):\n",
    "                orf_start = orf[0] + start\n",
    "                orf_end = orf[1] + start\n",
    "                orf_tag = tag + '.' + str(i)\n",
    "                orf_seq = seq[orf[0]:orf[1]]\n",
    "                \n",
    "                if len(orf_seq) < l_min: continue\n",
    "                \n",
    "                f = [SeqFeature(FeatureLocation(orf_start, orf_end, strand), type=\"CDS\")]\n",
    "                r = SeqRecord(orf_seq.seq, name=orf_tag, id=prokaryote_id, features=f)\n",
    "                cds.append(r)\n",
    "                \n",
    "                if i == 0: genes.append(r)\n",
    "    \n",
    "    # Find interregions\n",
    "    interregions = get_interregions(genes)\n",
    "    \n",
    "    # Get negative ORFs(longest ORF in interregion)\n",
    "    ## interregion (intr)\n",
    "    ## longest orf (long)\n",
    "    \n",
    "    ncs = []\n",
    "    for interregion in interregions:\n",
    "        feature = interregion.features[0]\n",
    "        \n",
    "        tag    = interregion.name\n",
    "        start  = feature.location.start.position\n",
    "        end    = feature.location.end.position\n",
    "        strand = feature.strand\n",
    "        \n",
    "        intr_seq = dna[0][start:end] if strand == 1 else dna[1][::-1][start:end][::-1]\n",
    "        \n",
    "        long_orf = longest_orf(intr_seq.seq)\n",
    "        if not long_orf: continue\n",
    "        \n",
    "        long_seq = intr_seq[long_orf[0]:long_orf[1]]\n",
    "        long_start = start + long_orf[0]\n",
    "\n",
    "        # check to see if noncoding ORF is larger than l_min (smallest coding sequence length) \n",
    "        if len(long_seq) < l_min: continue\n",
    "        \n",
    "        orfs = orf_finder(long_seq.seq)\n",
    "        \n",
    "        for i, orf in enumerate(orfs):\n",
    "            if orf[1]-orf[0] < l_min: continue\n",
    "            \n",
    "            orf_start = long_start + orf[0]\n",
    "            orf_end = long_start + orf[1]\n",
    "            orf_tag = tag + '.' + str(i)\n",
    "            orf_seq = long_seq[orf[0]:orf[1]]\n",
    "            \n",
    "            f = [SeqFeature(FeatureLocation(orf_start, orf_end, strand), type=\"NCS\")]\n",
    "            r = SeqRecord(orf_seq.seq, name=orf_tag, id=prokaryote_id, features=f)\n",
    "            ncs.append(r)\n",
    "    \n",
    "    return cds, ncs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "prokaryote_id = 'NC_000917'\n",
    "\n",
    "# get seq_records (features) from GenBank file\n",
    "handle = Entrez.efetch(db=\"sequences\", id=prokaryote_id, rettype=\"gbwithparts\", retmode=\"text\")\n",
    "seq_record = SeqIO.read(handle, \"gb\")\n",
    "handle.close()\n",
    "\n",
    "# get full sequence from FASTA file\n",
    "handle = Entrez.efetch(db=\"sequences\", id=prokaryote_id, rettype=\"fasta\", retmode=\"text\")\n",
    "sequence = SeqIO.read(handle, \"fasta\")\n",
    "handle.close()\n",
    "\n",
    "dna = [sequence, sequence.reverse_complement()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find ORFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_000917 ORFs extracted -- 7.5356 seconds\n",
      "52161 CDS ORFs\n",
      "43505 NCS ORFs\n",
      "95666 total ORFs\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "start_time = time.time()\n",
    "\n",
    "_cds, _ncs = preprocess_genome(seq_record, dna)\n",
    "\n",
    "print('%s ORFs extracted -- %s seconds' % (prokaryote_id, round(time.time() - start_time, 4)))\n",
    "print(f'{len(_cds)} CDS ORFs')\n",
    "print(f'{len(_ncs)} NCS ORFs')\n",
    "print(f'{len(_cds)+len(_ncs)} total ORFs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AF_NC00000.0 GTG TAA 4785 5025\n",
      "AF_NC00000.1 TTG TAA 4818 5025\n",
      "AF_NC00000.2 CTG TAA 4821 5025\n",
      "AF_NC00000.3 TTG TAA 4836 5025\n",
      "AF_NC00000.4 GTG TAA 4887 5025\n",
      "AF_NC00018.0 ATG TGA 25242 30024\n",
      "AF_NC00018.1 CTG TGA 25275 30024\n",
      "AF_NC00018.2 CTG TGA 25305 30024\n",
      "AF_NC00018.3 TTG TGA 25308 30024\n",
      "AF_NC00018.4 CTG TGA 25323 30024\n"
     ]
    }
   ],
   "source": [
    "# Check that noncoding ORFs start and end with correct codons\n",
    "for neg in _ncs[0:10]:\n",
    "    print(neg.name, neg.seq[0:3], neg.seq[-3:], neg.features[0].location.start.position, neg.features[0].location.end.position)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract Monocodon (Tri) and Dicodon (Hex) Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Produces all tri/hex codon permutations 4^n (ex. AAA, AAC,...,TTT for n=3) \n",
    "def permutations_with_replacement(n):\n",
    "    for i in itertools.product((\"A\", \"C\", \"G\", \"T\"), repeat=n):\n",
    "        yield i\n",
    "\n",
    "# Function to create codon:idx dictionary. Converts codons to integers (ex. {AAA:0, AAC:1, ..., TTT:63} for n=3)\n",
    "def populate_codon_idx_dict(nbases=3):\n",
    "    codon_idx_dict = {}\n",
    "    count = 0\n",
    "    for i in permutations_with_replacement(nbases):\n",
    "        key = \"\".join(i) #codon or dicodon sequence\n",
    "        codon_idx_dict[key] = count\n",
    "        count += 1\n",
    "    return codon_idx_dict\n",
    "\n",
    "\n",
    "TRICODON_IDX_DICT = populate_codon_idx_dict(nbases=3)\n",
    "HEXCODON_IDX_DICT = populate_codon_idx_dict(nbases=6)\n",
    "\n",
    "\n",
    "# Parse dna sequence into a list of codons separated by nbases\n",
    "# Input: Bio.Seq.Seq\n",
    "'''\n",
    "Ex. Input  - AAATTTGGG\n",
    "    Output - ['AAA','TTT','GGG'] if nbases=3\n",
    "           = ['AAATTT','TTTGGG'] if nbases=6\n",
    "'''\n",
    "def codon_parser(seq, nbases):\n",
    "    codon_seq = [seq[i:i+nbases] for i in range(0,len(seq),3) if (i+nbases)<len(seq)]\n",
    "    return codon_seq\n",
    "\n",
    "# Returns count of tri/hex codons in list form\n",
    "# Input: Bio.Seq.Seq\n",
    "def codon_feature(seq, nbases):      \n",
    "    codon_idx_dict = TRICODON_IDX_DICT if nbases==3 else HEXCODON_IDX_DICT\n",
    "    frame = [0]*len(codon_idx_dict)\n",
    "    \n",
    "    parsed_codons = codon_parser(seq, nbases)\n",
    "    for codon in parsed_codons:\n",
    "        frame[codon_idx_dict[codon]] += 1\n",
    "    frame[codon_idx_dict[parsed_codons[-1]]] += 1 # add last codon\n",
    "    \n",
    "    return frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "95666"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_orfs = _cds + _ncs\n",
    "len(_orfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Codon features extracted -- 496.2605 seconds\n"
     ]
    }
   ],
   "source": [
    "tri_temp = []\n",
    "hex_temp = []\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "for record in _orfs:\n",
    "    name = record.name\n",
    "    seq = record.seq\n",
    "    \n",
    "    t_frame = codon_feature(seq, 3)\n",
    "    tri_temp.append([name] + t_frame)\n",
    "    \n",
    "    h_frame = codon_feature(seq, 6)\n",
    "    hex_temp.append([name] + h_frame)\n",
    "    \n",
    "print('Codon features extracted -- %s seconds' % (round(time.time() - start_time, 4)))\n",
    "\n",
    "_tri = pd.DataFrame(tri_temp).reset_index(drop=True)\n",
    "_hex = pd.DataFrame(hex_temp).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>55</th>\n",
       "      <th>56</th>\n",
       "      <th>57</th>\n",
       "      <th>58</th>\n",
       "      <th>59</th>\n",
       "      <th>60</th>\n",
       "      <th>61</th>\n",
       "      <th>62</th>\n",
       "      <th>63</th>\n",
       "      <th>64</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AF_RS00005.0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AF_RS00005.1</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AF_RS00005.2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AF_RS00005.3</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AF_RS00005.4</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             0   1   2   3   4   5   6   7   8   9   ...  55  56  57  58  59  \\\n",
       "0  AF_RS00005.0   6   0   2   6   2   1   4   4   4  ...   0   1   0   2   2   \n",
       "1  AF_RS00005.1   5   0   2   6   2   1   3   4   4  ...   0   1   0   2   2   \n",
       "2  AF_RS00005.2   5   0   2   6   2   1   3   4   4  ...   0   1   0   2   2   \n",
       "3  AF_RS00005.3   5   0   2   6   2   1   3   4   4  ...   0   1   0   2   2   \n",
       "4  AF_RS00005.4   5   0   2   6   2   1   3   4   4  ...   0   1   0   2   2   \n",
       "\n",
       "   60  61  62  63  64  \n",
       "0   2   1   2   4   2  \n",
       "1   2   1   2   4   2  \n",
       "2   2   1   2   4   2  \n",
       "3   2   1   2   4   2  \n",
       "4   2   1   2   4   2  \n",
       "\n",
       "[5 rows x 65 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_tri.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>4087</th>\n",
       "      <th>4088</th>\n",
       "      <th>4089</th>\n",
       "      <th>4090</th>\n",
       "      <th>4091</th>\n",
       "      <th>4092</th>\n",
       "      <th>4093</th>\n",
       "      <th>4094</th>\n",
       "      <th>4095</th>\n",
       "      <th>4096</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AF_RS00005.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AF_RS00005.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AF_RS00005.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AF_RS00005.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AF_RS00005.4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 4097 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0     1     2     3     4     5     6     7     8     9     ...  \\\n",
       "0  AF_RS00005.0     0     0     0     0     0     0     0     1     0  ...   \n",
       "1  AF_RS00005.1     0     0     0     0     0     0     0     1     0  ...   \n",
       "2  AF_RS00005.2     0     0     0     0     0     0     0     1     0  ...   \n",
       "3  AF_RS00005.3     0     0     0     0     0     0     0     1     0  ...   \n",
       "4  AF_RS00005.4     0     0     0     0     0     0     0     1     0  ...   \n",
       "\n",
       "   4087  4088  4089  4090  4091  4092  4093  4094  4095  4096  \n",
       "0     0     0     0     0     0     0     0     0     0     0  \n",
       "1     0     0     0     0     0     0     0     0     0     0  \n",
       "2     0     0     0     0     0     0     0     0     0     0  \n",
       "3     0     0     0     0     0     0     0     0     0     0  \n",
       "4     0     0     0     0     0     0     0     0     0     0  \n",
       "\n",
       "[5 rows x 4097 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_hex.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract Translation Initiation Sites (TIS)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on the Gene Prediction paper by Hoff et. al, translation initiation sites (TIS) for the algorithm consist of up- and downstream regions centered at a start codon. They use a 60 basepair window at window position 31. They encode the trinucleotide occurrences in that window to yield binary indicator vectors. In each of its 3712 dimensions (64 trinucleotides × 58 positions), a vector indicates whether a certain trinucleotide occurs at a particular window position."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "60\n"
     ]
    }
   ],
   "source": [
    "# First we need to isolate the 60 basepair window with the start codon at its center (test case = cds[0])\n",
    "\n",
    "strand = cds[0].features[0].strand\n",
    "start  = cds[0].features[0].location.start.position\n",
    "\n",
    "# find window (-/+ 30 bp from start position)\n",
    "seq60 = (dna[0][start-30:start+30] if strand == 1 else dna[1][::-1][start-30:start+30][::-1]).seq\n",
    "print(len(seq60))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(64, 58)"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from tensorflow.keras.utils import to_categorical\n",
    "\n",
    "# find the trinucleotide index for each trinucleotide in the sequence (slide by 1 bp each time)\n",
    "frame = [TRICODON_IDX_DICT[seq60[i:i+3]] for i in range(0,len(seq60)-2)]\n",
    "\n",
    "# then convert the list into a 64x58 dimension frame to encode the trinucleotide occurances\n",
    "tis0 = to_categorical(frame, num_classes=64).astype(int).T\n",
    "tis0.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can visualize the frame using a heatmap - each dark square indicates the occurance of a trinucleotide at a position."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7ff5e3f958b0>"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAoUAAAHkCAYAAABIaVW2AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nOzdf7hdZX3n/fcnaa1EUCyDoUUiSFGU2DJRwKdQFBwVHYpEUHKondJagw72cmx9xFH7AK2V2ospKGXg2bWOwRlJrBhFoWoVEXkI6CGDSlAjggLiDwStFrWAfp8/9jrM5nhOzl4ne599svN+Xde62Ov+sda9EtDvda91f+9UFZIkSdq5LRn1ACRJkjR6BoWSJEkyKJQkSZJBoSRJkjAolCRJEgaFkiRJwqBQkiRph5LkXUm+m+SmWeqT5B1JbknyhSSr+rmuQaEkSdKO5d3AMduofwFwQHOsBS7s56IGhZIkSTuQqroauHcbTV4EXFxd1wG7J/m1ua77S4Ma4GBtdZsVSYvWLivOmLPNT24/awFGIu3MnpRR3n2XFRNDi1V+esf6U+nO8E3pVFWnxSX2Bu7oOb+zKfvWtjq1milMsjpJJTlwWvm/b8qfP0OfPZM8kOTUNveSJEnaGVVVp6qe0XO0CQgBZgqY5wxi274+ngCuAdbMUj4xQ5+XANfNUidJkrTDSZYM7RiAO4F9es4fD9w1V6e+75xkV+Bw4OX0BIVJApwInAI8L8kjp3WdAP4MeHySvfu9nyRJkublMuA/NauQnwn8S1Vt89UxtJspPB74aFVtBe7tWd58OHBbVX0NuAp44VSHJPsAe1XVZ4H3ASfNdvEka5NMJpnsdDa0GJYkSdLCCkuGdsx57+QSYBPw5CR3Jnl5klcmeWXT5ArgVuAW4O+B/9zPM7VZaDIBnNf8Xt+cb27+ub6n/PeBDzTna+gGg1N1/wD87UwXb96XN+/MXWgiSZI0k6ra5id5VVXAaW2v21dQmGQP4GhgZZIClgKV5A3ACcBxSd5E98PGPZLsVlU/ohswLk/ye82lfj3JAVX11bYDlSRJWiwG9O3fotLvTOGJdPPdPLSCOMmngTcDn6+q5/eUrwOOT/JZ4FFVtXdP3Vl0Zw//chCDl6RRMN2MpHHUb5g7AWycVnYp8MxZyk/eRh9XIUuSpB3aIl99PL9n6r52Xmz8plCSJG3LaJNX77bfHw4tVvnRbf9jJM+2SHc0kSRJWry6GfnGi0GhJElSa+O30GT8nkiSJEmtOVMoSZLU0s6ckkaS1MIuK86Ys42pbSQtJq3C3CSrk1SSA6eV//um/PnTyg9I8pEkX0tyQ5JPJTlyEAOXJEkalXFMSdP2zhPANXQTUM9U/lAOwiSPBC4HOlW1f1U9HfgT4InzH64kSZKGoe/Xx0l2BQ4HjgIuA85sykN3x5PnAp9J8siq+inwe8Cmqrps6hpVdRNw08BGL0mSNAIZw7W6bZ7oeOCjVbUVuDfJqqb8cOC2qvoacBXwwqb8IGBzvxdPsjbJZJLJTmdDi2FJkiQtrHF8fdxmockEcF7ze31zvrn55/qe8t8HPjC9c5KNwAHA1qp68fT6quoAne6ZO5pIkiQtpL6CwiR7AEcDK5MUsBSoJG8ATgCOS/ImIMAeSXYDtgAPLSqpqtVJngGcM+BnkCRJWlA7c0qaE4GLq+rUqYIknwbeDHy+qp7fU76O7qvm9wL/NclxPd8VLhvMsHduprrYsfj3tXPy71TSjqbfMHcC2Dit7FLgmbOUn1xVPwGOBV6Z5NYkm+gGkW/ZjvFKkiSN3E77TWFVPXuGsncA75ih/DK6q5Opqi/zfxaeSJIkaZFyRxNJkqSWQkY9hIEzKJQkSWppHBeajN8TSZIkqTVnCiVJkloax5lCg8IdkKkudiz+fUmSdgStw9wkq5NUkgN7yg5NcnWSryT5cpJ3JlnWU/+hJiWNJEnSDm8cU9LM584TwDXAGoAky4F/BE6vqicDTwE+CuzW1O8OrAJ2T7LfIAYtSZKkwWoVFCbZFTgceDlNUAicBqyrqk0A1fX+qvpOU38C8GG6+yKvQZIkaYe3ZIjHaLS98/HAR6tqK3BvklXASuCGbfSZAC5pjonZGiVZm2QyyWSns6HlsCRJkhbOOL4+brvQZAI4r/m9nm0EefDQq+XfAK6pqkryYJKVVXXT9LZV1QE63bOt1XJckiRJ2g59B4VJ9gCOBlYmKWApUMA64OnAh2bodhLwWOC2JACPpvsK+c3bN2xJkqTRGceUNG2e6ETg4qp6QlXtW1X7ALcBnwD+IMlhUw2TvCzJXnRnEo9p2u9LN3j0u0JJkqRFpk1QOAFsnFZ2Kd0gbw1wTpOS5kvA7wC/CqwArptqXFW3AT/sDSAlSZJ2NGHJ0I5R6fv1cVU9e4ayd/Sc/s4M3faeoc+qfu8pSZKkheGOJpIkSS2N4zeFBoWSJEktNQtox8r4hbmSJElqzZlCSZKklnx9LEkaW7usOGPONj+5/awFGImkUWi79/HqJJXkwJ6yQ5Nc3aSj+XKSdyZZ1tQdk+SzTfmNSTYkWTHoh5AkSVpI45iSpu2dJ4BraBJQN9vY/SNwelU9GXgK8FFgtyQrgfOBP6iqA6vqYOB/AfsOaOySJEkakDbb3O0KHA4cBVwGnAmcBqyrqk0AVVXA+5v25wBvraovTV2jqi4b2MglSZJGZBy/KWzzRMcDH62qrcC9SVYBK4EbZml/ELC534snWZtkMslkp7OhxbAkSZIWVrJkaMeotFloMgGc1/xe35z3JckewCeBZUCnqs6Z3qaqOkCne7a1WoxLkiRJ26mvoLAJ6o4GViYpYClQwDrg6cCHZui2BVgFfL6q7gEOTvI6YNdBDFySJGlURrkgZFj6nSk8Ebi4qk6dKkjyaeATwMVJLq+q65vylzXlfwNsTHJdz3eFywY3dEnSIJluRtq59RsUTgB/Pa3sUrqrkNcA5yR5HPBz4GrgA1X17SSvoRs07gbcA9wOzJ0IS5IkaTEbw4UmfQWFVfXsGcre0XP6O7P0uxy4fF4jkyRJ0oJxRxNJkqSWxjEljUGhJElSS0lGPYSBG78wV5IkSa05UyhJktTSzpySRtJOZJcVcycJMH3J9hvUn7N/X5IGoVWYm2R1kkpyYE/ZoUmuSvLVJJuTXJ7kadP6fT7JJYMatCRJ0iiN4zZ3be88AVxDNzchSZYD7wPeWFUHVNUq4Gxg/6kOSZ7S3OfIJI8ayKglSZI0UH0HhUl2BQ4HXk4TFAKvBtZV1bVT7arqmqr6YE/Xk4H3AB8HjtvuEUuSJI1aMrxjRNrMFB4PfLSqtgL3JlkFHARsnqPfScAG4BK6M40zSrI2yWSSyU5nQ4thSZIkLbAlQzxGpM1CkwngvOb3emYI8JJcDzwa+HhVvSbJIcDdVfWNJHcC70ry2Kr6/vS+VdUBOt2zrdXqKSRJkrRd+goKk+wBHA2sTFLAUqCAdcAq4EMAVXVYkhOBY5uuE8CBSb7enD8aOAF456AeQJIkacGNYfLqfmcKTwQurqpTpwqSfJrud4L/M8nHer4rXNbULwFeAvxmVX2zKTsKeDMGhdKiZvqShTGoP2f/vrafaX2k/oPCCeCvp5VdSncRyUnA25LsDXwX+B7wF8CRwDenAsLG1cBTk/xaVX1ru0YuSZI0KjvrTGFVPXuGsnf0nD5rlq7PnNbnZ8Cv9Ts4SZIkLQx3NJEkSWpr/Ha5MyiUJElqq8bw9fEYxrmSJElqy5lCSZKktsZvotCgUJIk080sDFP/LG5t9j5enaSSHNhTdmiSq5J8NcnmJJcneVpP/cuSfCHJliSfT/LOJLsP+iEkSZIW1JIM7xjVI7VoOwFcA6wBSLIceB/wxqo6oKpWAWcD+zf1xwCvBV5QVQfR3fnkWmD54IYvSZKkQeh3m7tdgcOBo4DLgDOBVwPrenYyoaqu6en2JuB1U8mrmxyF7xrMsCVJkkZoJ159fDzw0araCtybZBVwELB5G33mqn+YJGuTTCaZ7HQ29NtNkiRp4WWIx4i02ebuvOb3+ub8YZJcDzwa+HhVvWZa3dOA9wC70X3d/AtRX1V1gE73bGv1OS5JkiQNwJxBYZI9gKOBlUkKWAoUsI7ud4IfAqiqw5KcCBzbdN3S1H+qqr4IHJzk74BdBv4UkiRJC2mEC0KGpZ+ZwhOBi6vq1KmCJJ8GPg78zyQf6/mucFlPv7OBc5K8qKrubMoMCCVJ2kmZbmZx6yconAD+elrZpcDJwEnA25LsDXwX+B7wFwBVdUWSPYF/SrIU+AFwE/CxAY1dkiRpNMZwocmcQWFVPXuGsnf0nD5rG33X0X3NLEmSpEXMHU0kSZLaGr+JwlbJqyVJkjSmnCmUJElqayddfSxJkqRe4xcTGhRKkqRt22XFGXO2Md3Mjm/e3xQmWZ2kkhyY5PokNya5Pcndze8bk+yb5JeSfC/J2YMcuCRJ0qhUMrRjVLZnpnACuAZYU1WHASQ5BXhGVb16qlGSFwJfAV6a5I1V5RZ2kiRJi8y8ZgqT7AocDrwcWDNH8wng7cDtwDPncz9JkqRFZUmGd8whyTFJvpLkliRvmKH+MUk+nOTzSbYk+cO+HmkefwwAxwMfraqtwL1JVs0y6F2A5wAfAS6hGyDOKMnaJJNJJjudDfMcliRJ0vhqdom7AHgB8FRgIslTpzU7Dbi5qn4LeDbw35I8Yq5rz/f18QRwXvN7fXO+eYZ2xwKfqqofJ7kU+PMkr62qn01vWFUdoNM92+orZkmStHiN7tO/Q4FbqupWgCTrgRcBN/e0KWC3JAF2Be4FHpzrwq2DwiR7AEcDK5MUsBSoJK+fofkEcHiSrzfnewBHAZ9oe19JkqRFY4gLQpKsBdb2FHWayTOAvYE7euruBA6bdom/Ay4D7gJ2A06qqp/Pdd/5zBSeCFxcVadOFST5NHBEb6Mkj27K9qmqf2vK/pBuoGhQKEnSDsJ0Mwvr4W9Pf8FM0ej0N6zPB26kO4m3P/DPST5TVT/c1n3n803hBLBxWtmlwMnTyl4MXDkVEDY+BByX5FfmcV9JkqTFYXQLTe4E9uk5fzzdGcFefwh8oLpuAW4DDpzrwq1nCqvq2TOUvaPn9N1N2bunfve0uxfYs+09JUmSBMDngAOS7Ad8k24WmOkTc7fTXej7mSTLgScDt851YXc0kSRJamtEC02q6sEkrwY+Rnddx7uqakuSVzb1FwF/Cbw7yRebkZ5eVd+b69oGhZIkSTuQqroCuGJa2UU9v+8Cntf2ugaFkiRJbY1wO7phMSiUJElqy6BQGg+7rDhjzjYLmYJhsY1nR+SfoSRtn/nufbw6SSU5MMn1SW5McnuSu5vfNybZN8muSS5M8rUk/zvJDUleMeiHkCRJWlBLhniMyPZsc3cNsKaqDgNIcgrwjKp69VSjZuuVW4EDqurnSfYE/mj7hixJkqRBax2PJtkVOBx4Od3cOLO125/u/nxvntpaparurqq3zXOskiRJi0MyvGNE5jNJeTzw0araCtybZNUs7Q4CPt/PXnvQ3ecvyWSSyU5nwzyGJUmSpPmaz+vjCeC85vf65nzzXJ2SvAl4CfC4qvr16fUP3+dv6/Q9/CRJkhaP8Vt83C4oTLIH3c2VVyYpupm0K8nrZ2h+M/BbSZZU1c+r6q+Av0ryr9s9akmSpBGqufco3uG0nSk8Ebi4qk6dKkjyaeCI6Q2r6pYkk8Bbkvx5Vf0sySMZUGxt+olt889n2xbbsy+28QzKQv57OK5/huPK/42SFp+23xROABunlV3KL27EPOWPgT2AW5LcAHwCOL3lPSVJkhaXMVxo0mqmsKqePUPZO3pO3z2t7ofAqUiSJGlRc0cTSZKktsbvk8JR5s2WJEnSYuFMoSRJUluuPpYkSdIoF4QMS6oWY55ok1dLkqRtedJIo7L9/9OGocUqX7v4pJE8W6tvCpPslWR9kq8luTnJFUme1NS9NslPkzxmhn6rk1SSAwc1cEmSpJHJEI8R6TsoTBK6OQqvqqr9q+qpwBuB5U2TCeBzwOoZuk8A1wBrtm+4kiRJGoY2M4VHAQ9U1UVTBVV1Y1V9Jsn+wK7Am+kGgA9JsitwOPByDAolSdI4WJLhHaN6pBZtVwI3zFI3AVwCfAZ4cpLH9dQdD3y0qrYC9yZZNdMFkqxNMplkstPZ0GJYkiRJ2l6DWn28BlhdVT9P8gHgJcAFTd0EcF7ze31zvnn6BaqqA3S6Zy40kSRJi9hOnpJmC3Di9MIkvwkcAPxz97NDHgHcClyQZA/gaGBlkgKWApXk9bU4lz1LkiTNqcYvJmz1+vhK4FeSvGKqIMkhwNuBM6tq3+b4dWDvJE+gG0ReXFVPaOr2AW4DjhjgM0iSJGk79R0UNjN7q4HnNilptgBnAs+muyq510a6r5QnZqi7FDh5nuOVJEkavTFcaNLqm8Kqugt4aR/t/rT5+bYZ6t7R5p6SJEkaPre5kyRJamsMt7lrtaOJJEmSxpMzhZIkSW3t5ClpJEmSBGP5rtWgUJI01nZZccacbX5y+1kLMBJpcWsV5ybZK8n6JiXNzUmuSPKkpu61SX6a5DHT+hyT5LNJvpzkxiQbkqwY5ENIkiQtqGR4x4j0HRSmu13JRuCqqtq/qp4KvBFY3jSZAD5HN5fhVJ+VwPnAH1TVgVV1MPC/gH0HM3xJkiQNQpvXx0cBD1TVRVMFVXUjQJL9gV2B/5tuoPjupsnpwFur6ks9fS7bzjFLkiSN1hguNGnz+nglcMMsdRPAJcBngCcneVxTfhCwuZ+LJ1mbZDLJZKezocWwJEmStL0GtdBkDbC6qn6e5APAS4ALehsk2QP4JLAM6FTVOb31VdUBOt2zrTWgcUmSJA1cjWHy6jZB4RbgxOmFSX4TOAD45+5nhzwCuJVuULgFWAV8vqruAQ5O8jq6r5olSZJ2TDt5SporgbcmeUVV/T1AkkOAvwHOrKqzpxomuS3JE5q6jUmu6/mucNmAxi5J0pxMNyP1p++gsKoqyWrgvCRvAH4KfB14NvCqac03Amuq6m1JXgNcnGQ34B7gdmDupFGSJEmL1RguNGn1TWFV3QW8tI92f9rz+3Lg8vZDkyRJ0kJxRxNJkqS2xnChyRh+JilJkqS2nCmUJElqa2f/plCSJEnA+MWEvj6WJElSy6AwyV5J1if5WpKbk1yR5EnNcUWSW5J8Kcn7kizv6ff2JN9MYhAqSZJ2eLUkQztGpe8gLd3tSjYCV1XV/lX1VOCNwHK6KWcurKrfqKqnABcCezb9lgCrgTuAIwc8fkmSJA1Am28KjwIeqKqLpgqq6sYkfwRsqqoP95R/alq/m4ANwARw1XaNWJIkadTGcKFJm9e5K4EbWpRPmQAuoTvLeGySX56pUZK1SSaTTHY6G1oMS5IkSdtrqKuPkzwCeCHw2qr6UZLrgecxww4nVdUBOt2zrTXMcUmSJG2XMUxe3SYo3AKcOEv5s2bpcwzwGOCL3U8SWQb8GLe9kyRJO7IxXDrb5pGuBH4lySumCpIcAtwC/HaS/9hTfkySp9F9dfzHVbVvVe0L7Ac8L8mygYxekiRJA9F3UFhVRXcV8XOblDRbgDOBu4BjgT9J8tUkNwOnAD8Enk/PrGBV3QdcA/zuoB5AkiRpwSXDO0ak1TeFVXUX8NJZqo+ZoexXZ7jGi9vcU5IkScPnNneSJElt7eQpaSRJkjSmnCmUJElqy5lCSZIkjSNnCiVJklqqMUxe3fdMYZK9kqxv0tHcnOSKJE9qjiuS3JLkS0nel2R50+fQJFc1qWo2J7m8yV8oSZK041oyxGNE+popTHc7ko3Auqpa05QdDCwH3gX8aVV9uCk/Ctiz2cHkfcDJVXVtU3cEsD/wxQE/hyRJkrZDv6+PjwIeqKqLpgqq6sYkfwRsmgoIm/JPAST5S7pB5LU9ddcMZtiSJEkjtBO/Pl4J3NCiHOAgYHO/A0myNslkkslOZ0O/3SRJkjQAC7bQJMn1wKOBj1fVa6bXV1UH6HTPttZCjUuSJKm1nTglzRbg6S3Kp+pWTZ1U1WHAnwOPaTNASZIkDV+/QeGVwK8kecVUQZJDgFuA307yH3vKj2lWGF8AnJLkt3uus2wAY5YkSRqtJRneMapH6qdRVRWwGnhuk5JmC3AmcBdwLPAnTdqZm4FTgO9W1beBk4Czm3Q11wInAn83+MeQJElaQBniMSJ9f1NYVXcBL52l+phZ+lwHPGse45IkSdICckcTSZKklmonXmgiSZKkMeZMoSRJUls7cfJqSZIkjTFnCiVJfdtlxRlztvnJ7WctwEikEduZvylMsleS9U1KmpuTXJHkSUkOSPKRpvyGJJ9KcuS0vh9Ksmnww5ckSRqBMUxJ01dQmCTARuCqqtq/qp4KvBFYDlwOdJrypwN/Ajyxp+/udHc22T3JfoN+AEmSJG2/fl8fHwU8UFUXTRVU1Y1JXg5sqqrLespvAm7q6XsC8GHgO8Aa4OztHrUkSdIILRnDVRn9PtJK4IYZyg8CNs/RdwK4pDkmZmuUZG2SySSTnc6GPoclSZKkQRjoQpMkG4EDgK1V9eIky4HfAK6pqkryYJKVzWziw1RVB+h0z7bWIMclSZI0SGOYkabvmcItwNNnKV81dVJVq+nuffyrTdFJwGOB25J8HdiX7itkSZIkLSL9zhReCbw1ySuq6u8BkhwC3AL81yTH9XxXuKyn3wRwTFVtavrsB/wz8OaBjF6StKBMNyN17bQzhVVVwGrguU3qmS3AmcBdwLHAK5Pc2qSdeTPwliT7AiuA63qucxvwwySHDfIhJEmSFlKSoR2j0vc3hVV1F/DSWapfOEv53jNcZ9VMDSVJkjS3JMcAbweWAu+sqr+eoc2zgfOAXwa+V1XPmuu67mgiSZLU0qgm9JIsBS4AngvcCXwuyWVVdXNPm92B/073E77bkzyun2uPYZYdSZKksXUocEtV3VpV9wPrgRdNa3My8IGquh2gqr7bz4UNCiVJklpKhnn8n9zNzbG259Z7A3f0nN/JL36u9yTgsUmuarYg/k/9PJOvjyVJkhaRh+du/gUzvbient/5l+imEnwOsAuwKcl1VbV1W/c1KJQkSWopo3vXeiewT8/54+lmg5ne5ntVdR9wX5Krgd8CthkU9v1ISfZKsr5JSXNzkiuSPCnJAUk+0pTfkORTSY7s6XdMks8m+XKSG5NsSLKi3/tKkiQtNsN8fTyHzwEHJNkvySPobgpy2bQ2HwJ+J8kvJVkGHAZ8aa4L9zVTmG7SnI3Auqpa05QdDCwH/gF43VTy6iQrgWcAVze/zweOq6ovNfXH0d3Z5PZ+7i1JkqSuqnowyauBj9FNSfOuqtqS5JVN/UVV9aUkHwW+APycbtqaX9hieLp081LP0Sg5Gjizqo6cVv5y4Miq+oNZ+r0HuLKq/secN3kY9z6WJEnb8qSR7inylH+4emixypdefuRInq3f18crgRtmKD8I2LyNfnPVP6R3pU2ns6HPYUmSJGkQBrrQJMlG4ABga1W9eFrdHsAn6e6N3Kmqc3rrH77SxplCSZK0eO20ex8DW+gubZ6p/KFt66pqNXAK8KvT66vqnqo6mG7gt+s8xytJkqQh6DcovBL4lSSvmCpIcghwC3B4s3hkyrKe338DvCnJU2aplyRJ2uGMcPXx0PT1+riqKslq4LwkbwB+Cnwd+C/AscDfJjkP+A7wI+AtTb8vJnkNcHGS3YB76K46PmPQDyJJkrRQMobvj/v+prCq7gJeOkv1C7fR73Lg8pbjkiRJ0gJyRxNJkqSWRrijydCM4SNJkiSpLWcKJUmSWhrDTwqdKZQkSZIzhQtqlxVzL7r+ye1nLcBIJEk7A/9/Z3icKWwk2SvJ+iRfS3Jzkk8l+XGSG5Pcm+S25vcnmvb/Pkklef5ghy9JkrTwdto8hb3STcyzEVhXVWuasoOB3arqM0neDXykqt7f020CuKb558e2e9SSJEkaqPm8Pj4KeKCqLpoqqKobZ2vcBJEnAs8FPpPkkVX103ncV5IkaVFY4utjAFYCN7RofzhwW1V9DbiKWRJdJ1mbZDLJZKezYR7DkiRJ0nwtxEKTCWB983s98PvAB6Y3qqoO0Omeba0FGJckSdK8jONCk/kEhVvovg6eU5KlwAnAcUneBATYI8luVfWjedxbkiRJQzCfoPBK4K1JXlFVfw+Q5BBgWVV9elrb/wB8vqoeWnWcZB1wPPCeeY55h+Wyf0nSQvL/d4ZnHGcKW39TWFUFrAae26Sk2QKcCdw1Q/MJuiuVe10KnNz2vpIkSYtFlmRox6jM65vCqroLeOksdafM9Lun7DLgsvncV5IkScPhjiaSJEkt+fpYkiRJY8mZQkmSpJacKZQkSdJYcqZQkiSppXGcKWwVFCbZCzgPOAT4N+A7wGHAVmAF8C/N8b2q+g9JDgDOBZ4C/AD4IXBGVV09sCeQJElaYOO493HfQWGS0M05uK6q1jRlBwO7VdVnkrwb+EhVvb+peyRwOfC6Jg0NSVYCzwAMCiVJkhaRNjOFRwEPVNVFUwVVdeM22v8esGkqIGza3wTc1HqUkiRJi8g4vj5us9BkJXBDi/YHAZv7bZxkbZLJJJOdzoYWt5EkSdL2WrCFJkk2AgcAW6vqxdPrq6oDdLpnW2uhxiVJktRWxjB/S5tH2gI8vWX7VVMnVbUaOAX41RbXkCRJ0gJoM1N4JfDWJK+oqr8HSHIIsKyqPj1D+/cC/zXJcT3fFS7bvuFKkrS47bLijG3W/+T2sxZoJBqmcfymsO+gsKoqyWrgvCRvAH4KfB34L7O0/0mSY4G/TXIe3fQ1PwLest2jliRpEZorINT4yBhGha2+Kayqu4CXzlJ3ygxlXwZeOK+RSZIkacG4o4kkSVJLYzhR6N7HkiRJcqZQkiSpNWcKJUmSNJacKZQkSWppHGcK+w4KkywHzgWeCXwfuB/4m6ra2NS/HTgR2Keqfj6t72uBs4HlVfUvAxq7NHL9pJ8wJ9mOxb9TbQ//3dCOrK/Xx+km4/kgcHVVPbGqng6sAR7f1C8BVgN3AEfOcIkJ4HNNG0mSpB3akgzvGNkz9dnuaOD+qrpoqqCqvlFV5zenRwE3ARfSDQAfkmR/YFfgzcuoQHwAACAASURBVNPrJEmSdkQ7c1B4ELB5G/UTwCXARuDYJL88Q91ngCcnedx8BipJkqThmdfq4yQXJPl8ks8leQTdXUs+WFU/BK4HntfTfA2wvvnO8APAS2a55tokk0kmO50N8xmWJEnSgliSGtoxKv0uNNkCnDB1UlWnJfl3wCRwDPAY4IvNPoDLgB8Dlyf5TeAA4J+bukcAtwIXTL9BVXWATvds6+j+RCRJknZC/c4UXgk8MsmresqWNf+cAP64qvatqn2B/YDnJVnW1J05VVdVvw7sneQJAxq/JEnSghvHbwr7mimsqkpyPHBuktcDdwP3AWfQTVNzak/b+5JcA/wu3VfHL5h2uY1N+du2f/jSaJl+YvyM69+pqXYkzaXvPIVV9S26wdx062Zo++Lm5y98HFhVf9r36CRJkhahcdwSzh1NJEmSWhrlgpBhGcdAV5IkSS05UyhJktTSKBeEDIszhZIkSXKmUJIkqa1xnFUzKJSkRWxQqWRMNyNpLn0HukmWJ3lvkluT3JBkU5LVPfVvT/LNJEum9XtZki8k2dJsjffOJLsP8iEkSZIW0jgmr+4rKEx3j7oPAldX1ROr6ul0cxY+vqlfAqwG7gCO7Ol3DPBa4AVVdRCwCrgWWD7Ih5AkSVpISQ3tGJV+Xx8fDdxfVRdNFVTVN4Dzm9OjgJvoJqueAK5qyt8EvK6qvtn0+Rnwru0ftiRJkgap39fHBwGbt1E/AVxCdwu7Y5P8cp/9HpJkbZLJJJOdzi9shCJJkrRojOPr43ktNElyAXAEcD9wOPBC4LVV9aMk1wPPAy6f1udpwHuA3YA3VtXDIr+q6gCd7tnW8UsTLkmStIj1GxRuAU6YOqmq05L8O2ASOAZ4DPDF7qeHLAN+TDco3EL3O8JPVdUXgYOT/B2wy8CeQJIkaYHtzClprgTemuRVVXVhU7as+ecE8MdVdQlAkkcBtyVZBpwNnJPkRVV1Z9N+zoBwUCkYJA2P/50ujB3xz9B/N6QdU19BYVVVkuOBc5O8HrgbuA84AzgXOLWn7X1JrgF+t6o2JNkT+KckS4Ef0F2Q8rEBP4ckSdKCWTLCVcLD0vc3hVX1LbppaKZbN0PbF/f8XjdTG0mSpB2Vex9LkiRpLLnNnSRJUkvjOKs2js8kSZKklpwplCRJamkcvylclEGhqQqkxc//TjUb/92Qdkx9vz5OsjzJe5PcmuSGJJuSrG7qDk1ydZKvJPlyknc2eQqn+n4oyaZhPIAkSdJCW5Ia2jEqfc0UprtVyQeBdVV1clP2BOC4JMuBfwTWVNWmpu0JdLez+3GS3enuavKvSfarqtuG8SCSJEkLZRxfH/c7U3g0cH9VXTRVUFXfqKrzgdPoBoubmvKqqvdX1XeapicAHwbWM3OeQ0mSJI1Yv0HhQcDmWepWAjdso+8EcElzTMzWKMnaJJNJJjudDX0OS5IkaeEtGeIxKvNaaJLkAuAI4H7gjm20Ww78BnBNs1Xeg0lWVtVN09tWVQfodM+2jt/eMZIkSYtYvwHpFrrfBQJQVacBzwH2bOqePku/k4DHArcl+TqwL75CliRJO7hxXGjSb1B4JfDIJK/qKZtaXfx3wB8kOWyqIsnLkuxF93XxMVW1b1XtSzd4NCiUJElaZPp6fdy8+j0eODfJ64G7gfuA06vqO0nWAOckeRzwc+Bqut8grgCu67nObUl+mOSwqrp+0A8jSZK0EMZx9XHf3xRW1beYZZavWXn8OzNU7T1D21UztJMkSdphjGNQ6N7HkiRJWpzb3EmSJC1m4zirNo7PJEmSpJacKZQkSWpplKljhsWZQkmSpB1IkmOSfCXJLUnesI12hyT5WZIT+7lu30FhkuVJ3pvk1iQ3JNmUZHVTd2iSq5sBfjnJO5Ms6xn4Z5vyG5NsSLKi3/tKkiQtNksyvGNbkiwFLgBeADwVmEjy1FnavQ34WL/P1Nfr4yQBPgisq6qTm7InAMc1W9n9I7CmqjY1bU8AdkvyROB84Liq+lLT7zi6O5vc3u8gJUmSFpMRvmo9FLilqm4FSLIeeBFw87R2fwJcChzS74X7faajgfur6qKpgqr6RlWdD5xGN1jc1JRXVb2/qr4DnA68dSogbOovq6qr+x2gJEnSziTJ2iSTPcfanuq9gTt6zu9kWl7oJHsDq4GLaKHfoPAgujuUzGQlcMM8+j1M7x9Ap7Ohz2FJkiQtvGG+Pq6qTlU9o+fo9Nx6phfM01e9nEd317mftXmmea0+TnIBcARwPw+PVrfVZw/gk3T3TO5U1Tm99c0DNw+9dfyW9EiSJG2/O4F9es4fD9w1rc0zgPXdL/r4d8ALkzxYVR/c1oX7nSncAjy0PV1VnQY8B9izqXv6XP2q6p6qOphu4Ldrn/eVJEladJIa2jGHzwEHJNkvySPobkF8WW+Dqtqvqvatqn2B9wP/ea6AEPqfKbwSeGuSV1XVhU3Zsuaffwd8NsnlVXV99w8qLwM+AfwNsDHJdT3fFS5DkqQdzC4rzpizzU9uP2sBRtK12MajhVFVDyZ5Nd1VxUuBd1XVliSvbOpbfUfYq6+gsKoqyfHAuUleD9wN3Ef3ffV3kqwBzknyOODnwNXAB6rq20leA1ycZDfgHrqrjuf+N1mSJGmRmit1zDBV1RXAFdPKZgwGq+qUfq/b9zeFVfUtulOUM9VtAn5nlrrLgcv7vY8kSdJiN467f4zjM0mSJKkl9z6WJElqyb2PJUmSNJacKZQkSWpplAtNhsWgUJKkPiy29C6LbTza8fX9+jjJ8iTvTXJrkhuSbEqyuqk7NMlVSb6aZHOSy5M8bVr/zye5ZNAPIEmStNCGuc3dqPQ1U5juPikfBNZV1clN2ROA45IsB94HnFxV1zZ1RwD7A19szp9CNwA9Msmjquq+gT+JJEnSAlk66gEMQb+vj48G7u9NjFhV3wDOT/KXdIPFa3vqrpnW/2TgPcBTgOMAZwwlSZIWkX5fHx8EbJ5H3ZSTgA10g8GJmRokWZtkMslkp7Ohz2FJkiQtvCWpoR2jMq+FJkkuAI4A7gfumFZ3PfBo4ONV9ZokhwB3V9U3ktwJvCvJY6vq+739qqoDdLpnW8cv+Y8kSdIi1u9M4RZg1dRJVZ0GPAfYc4a6w4A/Bx7TFE0AByb5OvA1ugHjCds7cEmSpFEZx4Um/QaFVwKPTPKqnrJlzT8vAE5J8tvT65IsAV4C/GZV7VtV+wIvYpZXyJIkSRqNvl4fV1UlOR44N8nrgbuB+4DTq+rbSU4C3pZkb+C7wPeAvwCOBL5ZVd/sudzVwFOT/FpVfWuQDyNJkrQQdurk1U0At2aWuuuAZ83S9ZnT2v4M+LV+7ytJkrTYLB3DoNC9jyVJkuQ2d5IkSW2N4+tjZwolSZLkTKEkSVJbo0wyPSwGhZIkadHYZcUZfbX7ye3umDtofb0+TrI8yXuT3JrkhiSbkqxu6g5NclWSrybZnOTyJE/r6fuyJF9IsiXJ55O8M8nuw3ogSZKkYRvH5NVzzhQmCfBBYF1VndyUPQE4Lsly4H3AyVV1bVN3BLA/8MUkxwCvBV5QVd9MshT4A2A58INhPJAkSZLa6+f18dHA/VV10VRBVX0DOD/JX9INFq/tqbump++bgNdNJa9uchS+ayAjlyRJGpGlox7AEPTz+vggYPM86vqpf0iStUkmk0x2Ohv66SJJkjQSO+Xr4+mSXAAcAdwP3DGt7nrg0cDHq+o10+qeBrwH2A14Y1U9LPKrqg7Q6Z5tHb8lPZIkSYtYPzOFW4BVUydVdRrwHGDPGeoOA/4ceMz0vlX1xao6GPgnYJdBDF6SJGkUlqSGdoxKPzOFVwJvTfKqqrqwKVvW/PMC4PokH+v5rnBZT9+zgXOSvKiq7mzKDAglSdKMfnL7WaMewk5rzqCwqirJ8cC5SV4P3A3cB5xeVd9OchLwtiR7A98Fvgf8RdP3iiR7Av/UrDz+AXAT8LHhPI4kSdLwLR3Dbe76+qawqr4FrJml7jrgWdvouw5YN6/RSZIkaUG4o4kkSVJLo1wlPCwGhZIkSS2NY1DY1zZ3kiRJGm/OFEqSJLU0jjOFBoXSGNllxRlztjHdgyRpJvPZ0WQ5cC7wTOD7dHc2eTTwAPAIYD/gK03zt1TV+5N8CHhcVf1fAxm1JEnSCC0dYZLpYWkVFCYJ8EFgXVWd3JQ9ATiuqs5Psi/wkWbnkqk+u9Pd1eRfk+xXVbcNavCSJEkajLYLTY4G7q+qi6YKquobVXX+NvqcAHwYWM8suQ4lSZJ2JEuGeIxK23sfBGxu2WcCuKQ5JmZrlGRtkskkk53Ohpa3kCRJWjhLMrxjVLZroUmSC4Aj6M4eHjJD/XLgN4Brmu3yHkyysqpumt62qjpAp3u2dfxe1EuSJC1ibWcKt9D9PhCAqjoNeA6w5yztTwIeC9yW5OvAvvgKWZIk7eCcKYQrgbcmeVVVXdiULdtG+wngmKraBJBkP+CfgTe3HqmkOZluRpI0X62CwuYV8PHAuUleD9wN3AecPr1tsxJ5BXBdT//bkvwwyWFVdf32DFySJGlUdvqUNABV9S1meQVcVV8HVvb83nuGNquml0mSJGm03NFEkiSpJbe5kyRJ0lgGhaPMkShJkqRFwplCSZKklpwplCRJ0lhqNVPY7FByLvBM4PvA/cCjgQeARwD7AV9pmr+lqt6f5E+BtU2bnwOfBE6vqgcG8gSSJEkLbOkYzhT2HRQmCfBBYF1VndyUPQE4rqrOb/ISfqSqDu7p80rgecAzq+oHSR4B/CmwC90gUZIkSYtAm5nCo+nucXzRVEFVfQM4fxt93gQcWVU/aNrfD/z1fAYqSZK0WCwZw+TVbb4pPAjY3G/jJLsBu1bVbX22X5tkMslkp7OhxbAkSZIW1pIhHqMy79XHSS4AjqA7e3jITE2A6mn/fOBtwO7AyVV1bW/jquoAne7Z1vELvyVJkhaxNgHpFuChLeqq6jTgOcCeMzWuqh8C9yXZrzn/WPO94U10F6VIkiTtkJZkeMfInqlF2yuBRyZ5VU/Zsjn6nA1cmGR3eGixyiPbDVGSJEnD1vfr46qqJMcD5yZ5PXA3cB9w+ja6XUg3cLw+yb8B/wr8f8D/nv+QJUmSRmscU9KkajF+vuc3hZIkaVueNNKw7NPfumJoscqzfu2FI3k2t7mTJElqaRxT0hgUSpIkteTex5IkSRpLzhRKkiS1NI4zhQaFi8wuK86Ys81Pbj9rAUYiSZJ2Jq1eHyfZI8mNzfHtJN/sOV+e5IEkp87Q75eSfC/J2YMbuiRJ0miM4zZ3re5dVfdU1cHNziQXAef2nJ8AXAdMzND1ecBXgJc2CawlSZK0iAwyIJ0A/gx4fJK9Z6h7O3A78MwB3lOSJGnBJcM7RmUgQWGSfYC9quqzwPuAk3rqdqG7R/JHgEuYeSaRJGuTTCaZ7HQ2DGJYkiRJQ5EhHqMyqJnCNXSDQYD1PDzwOxb4VFX9GLgUWJ1k6fQLVFWnqp5RVc9Yu/ak6dWSJEkaokGtPp4Alif5veb815McUFVfbeoOT/L1pm4P4CjgEwO6tyRJ0oIaxxUS2x0UJnky8Kiq2run7CxgTZK3A0cA+1TVvzV1f0g3UDQonIHpZiRJ0igMYqZwAtg4rexSuq+R7wCunAoIGx8C/ibJr0wrlyRJ2iGM45Zw8w4Kq+rMbdR9AXhqc/ruaXX3AnvO976SJEkaPHc0kSRJaimpUQ9h4AwKJUmSWhrDdSZj+UpckiRJLTlTKEmS1JIpaSTt8HZZccacbUyNJKkt/7dlx9fq9XGSPZLc2BzfTvLNnvPlSR5Icuq0PsuTvDfJrUluSLIpyerBPoYkSdLC2em3uauqe6rq4Ko6GLgIOLfn/ATgOnq2uEsS4IPA1VX1xKp6Ot0t8R4/sCeQJEnSdhvkQpMJ4M+AxyeZ2t3kaOD+qrpoqlFVfaOqzh/gfSVJkhbUkgzvGNkzDeIiSfYB9qqqzwLvA05qqg4CNvd5jbVJJpNMdjobBjEsSZKkoRjH18eDWmiyhm4wCN3t7f4B+NvpjZJcQHcv5Pur6pDeuqrqAJ3u2dbxywgpSZK0iA3q9fEEcEqSrwOXAb+V5ABgC7BqqlFVnQY8B7e5kyRJO7BkeMfc984xSb6S5JYkb5ih/veSfKE5rk3yW/0803bPFCZ5MvCoqtq7p+wsurOHbwHemuRVVXVhU71se+8pbYtpEbZtR3x2/06l4RnUf1/+N7gwkiwFLgCeC9wJfC7JZVV1c0+z24BnVdX3k7yA7pvYw+a69iBmCieAjdPKLgUmqqqA44FnJbktyWeBdcDpA7ivJEnSSIzwm8JDgVuq6taqup/uZ3sv6m1QVddW1feb0+voM+vLvGcKq+rMbdR9AXhq8/tbdGcNJUmSNIcka4G1PUWdZu0FwN7AHT11d7LtWcCXA//Uz33d0USSJKmlYa4Sfvji275uPeMC3SRH0Q0Kj+jnvgaFkiRJLY0wn+CdwD49548H7preKMlvAu8EXlBV9/Rz4UEmr5YkSdJwfQ44IMl+SR5B9xO9y3obJFkBfAD4/ara2u+FnSmUJElqaVQThVX1YJJXAx8DlgLvqqotSV7Z1F8E/D/AHsB/7+44zINV9Yy5rp3uAuHFxuTVkqS5ma5oZ/akUW7+wVf/5SNDi1UOeMyxI3m2VjOFSfYAPtmc7gX8DLi7OX8v8EfAg035f6uqi5t+e9J93/3qqvp/BzBuSZKkkUnGb/6q1TeFVXVPVR1cVQcDFwHn9vx+DnBoVa0EjuThM6svoZsnZ2Iww5YkSdIgDWqhyRuB/1xVPwSoqn+pqnU99RPAnwGPT7L3TBeQJEnaUYwwefXQbHdQmGQ3YLeq+tos9fsAe1XVZ4H3ASfN0m5tkskkk53Ohu0dliRJkloYxOrjMEvSxMYausEgdLdi+Qfgb6c3eniiRheaSJKkxSsjXeYyHNsdFFbVD5Pcl+SJVXXrDE0mgOVJfq85//UkB1TVV7f33pIkSaMwjomeB5Wn8GzggiQnNUHio+nOEH4aeFRVPfQdYZKzmrq/HNC9JUk7KdPNSIMzqED3QuBTwOeS3EQ3GPwx3VnCjdPaXoqrkCVJ0g4sGd4xsmcyebUkSdrxjDZ59Tf+9cNDi1WesOvvLv7k1ZIkSRpt6phhGcfvJCVJktSSM4WSJEktmZJGkiRJvj6WJEnSeOp7pjDJHsAnm9O9gJ8Bdzfn7wX+CHiwKf9vVXVxkl8C/gJ4CXBf0/Yfq+qvBjB2SZKkkVgyhlOFfQeFVXUPcDBAkjOBf62qc5K8ElgNHNokrn4McHzT7S10A8inVdVPm32S/2yQD/D/t3fn0XaV5R3Hvz8TUggzAZkRRSiDy4YISDFVQCqQ0gDKWhJwwKGpLGQhXSgoy3LV1pZAC63QsO5CJQ5AKJIuCm2hhSKyApFLypAUisxTlEGGMCgRn/6x3wOb4xn2u885uTcnv0/WWXePz373c/a+ee+79363mZmZmfWuH/cUfgU4ICJeAIiI54EFkqYCfwbsGBG/SvNWAiN92KaZmZnZuBnChsLe7ilMLX8bRsT9LWa/E3gkVQSrxJoraUzS2Ojowl6KZWZmZmaZem0pFFCpR29JnwJOAqYB+0XEo+X5ETEKjBZjfqOJmZmZTVzS8FVVemopTJeMX5L0jhaz7wN2SK2JRMR3I2I68DwwqZftmpmZmY0nDfAzXvpxT+HfAOdL+mh60GQj4OiIGJX0beA8SX+eHjSZBEzpwzYnnPV2OKPrMq888rXVUBIzMzOzfP2oFM4HNgBulbQKWAX8XZp3OvANYJmklcArwALgiT5s18zMzGxc+I0mSUSMlIYDmJc+zcutAk5LHzMzMzOboPyaOzMzM7NMQ9hQ6NfcmZmZmZlbCs3MzMyyDWOrmiuFZmZmZpn8oIm15e5mho+7GRo+/k7NzNqr3PopaZqk29Pn55IeL41vKeliSQ9Iuk3SzZKObFr/H9I6w9jiamZmZmuV4eu+unJLYUQ8A0wHkDQCvBgRZ0sSsBhYEBHHpPlvA2Y31k0VwSOBR4H3Azf0qfxmZmZm1gf9uHx8IPBqRFzQmBARDwPfKi1zALAMWAjMwZVCMzMzW4NpCDul6cel3D2ApV2WmQNcAiwCDpO0TvMCkuZKGpM0Njq6sA/FMjMzM7Oq+v6giaTzgZkUrYd7S5oCzAJOjoiVkpYAHwKuLq8XEaPAaDF2b/S7XGZmZmb9MoyPSPSjUrgc+EhjJCJOkLQ5MJYmHQJsDNxV3H7IVOBlmiqFZmZmZmuO4bt83I9K4fXANyUdHxHz07SppflzgM9GxCUAktYHHpQ0NSJe7sP2zQbCXZMMH3+ntqZw90k2Hnpu+4yIAI4APiDpQUk/BRYAp0qaChxMqVUwIl4CbgL+tNdtm5mZmY0HDfDfeKnVUhgRI03jK4Cj2yy+WYv1P1xnu2ZmZmY2GH6jiZmZmVm24buncPgenTEzMzOzbG4pNDMzM8vkLmnMzMzMjGG8fOxKoZmZ2QTj7mZsPFSqFEqaBlyXRrcCXgOeSuMHA+cA+wLPAq8C8yJiUVp3H2AesC2wElgBnBYRd/VpH8zMzMxWq2F893GlSmFEPANMB5A0ArwYEWereEXJYmBBRByT5r8NmJ2GtwQuA46JiMVp2kxgJ8CVQjMzM7MJotfLxwdSvOP4gsaEiHgY+FYa/TxFhXFxaf5NPW7TzMzMbFwNY0thr4/O7AEs7WH+6yTNlTQmaWx0dGGPxTIzMzOzHH190ETS+cBMitbDvVvMXwJsBFwbESeV50XEKDBajN0b/SyXmZmZWX8NX5c0ve7RcmBGYyQiTgA+CGzRZv57ga8CG/e4XTMzM7NxI2lgn/HSa0vh9cA3JR0fEfPTtKml+ecDSyRdU7qvcCpmlm29Hc7ouoy7sTAzs7p6qhRGREg6AjhH0pcouql5CTg1zf+5pI8CZ0raFngSeBr4em/FNjMzMxtPw/egSXalMCJGmsZXAEd3WP4W4APZJTMzMzOz1cZvNDEzMzPL5C5pzMzMzGwouaXQzMzMLNvwtau5UmhmZmaWaRgvH7tSaGuUtblblmHdr7X5OzUzm0iyK4WSpgHXpdGtgNfSz7uAKWn4+fR5OiIOkrQnxevuDomIa/pRcDMzM7PxMp6dTA9KnS5pngGmA0gaAV6MiLMb8yVdBFwVEZeXVpsD3JR+ulJoZmZmNsEM/PKxiqr0UcAfAz+RtG5E/GrQ2zUzMzMbnOFrKVwdj868D3gwIu4HbgBmtVpI0lxJY5LGRkcXroZimZmZmVnD6njQZA5waRq+FPg4cEXzQhExCowWY/fGaiiXmZmZWS1ylzR5JE0CPgLMlnQ6RVvrNEkbRsTKQW7bzMzMbHCG7/LxoFsKDwLuiIiDGxMkLQCOAL4/4G33jbvMmDic5+Hj79Ta8e9es9Vr0G2fc4BFTdN+BBwz4O2amZmZDYykgX3GS08thREx0mLaca2GS9OuBK7sZbtmZmZm1l9+o4mZmZlZtuG7p3D4Hp0xMzMzs2xuKTQzMzPL5C5pzMzMzIxhvHysiInYT7Q7rzYzs+Hkrnb6ZZdxrZWt+u3tA6urrPOW6eOyb1kthZKmAdel0a2A19LPu4Apafj59Hk6Ig6StDNwDrAb8BzwAnBGRNzYlz0wMzMzW800hC2FWZXCiHgGmA4gaQR4MSLObsyXdBFwVURcnsbXBa4GTkld0SDpXcBegCuFZmZmZhPEoO8pPBa4uVEhBIiIZcCyAW/XzMzMbGDGs5PpQRn0ozN7AEurLChprqQxSWOjowsHXCwzMzMzK1utTx9LWgTsDNwbER8uz4uIUWC0GPODJmZmZjaRDV+XNIPeo+XAjMZIRBwJHAdsNuDtmpmZmVmGQbcUXgx8WdLs0n2FUwe8zddVeey/imHtGqBf+VmdhvW7MLO1h3+PDYe1/unjXBHxiqTDgL+XdC7wC2Al8FeD3K6ZmZnZYLlS+LqIGGkx7bgW0+4BZtXdjpmZmZkNnl9zZ2ZmZpbJXdKYmZmZ2VBypdDMzMws21sG+OlM0iGS/k/SfZJOazFfkv4xzb9T0oxWcVrtkZmZmZmtASRNAs4HDgV2B+ZI2r1psUMp+oXeGZgLzK8S25VCMzMzs0wa4L8u9gHui4gHIuJV4FLg8KZlDge+F4VbgE0kbd11pyJijfgAcx3HcRzHcRzHcRzHcYb9Q9G6N1b6zC3NOwq4sDT+ceC8pvWvAmaWxq8D9uq23TWppXCu4ziO4ziO4ziO4zjOsIuI0YjYq/QZLc1u1ZTY/HrgKsv8jjWpUmhmZma2tnsM2L40vh3wRI1lfocrhWZmZmZrjluBnSW9XdIU4GjgyqZlrgQ+kZ5C3hd4PiJWdAu8JnVePdp9EcdxHMdxHMdxHMdZS+OsFSLiN5I+D1wDTAK+ExHLJX0uzb8A+DeKt8ndB7wMfKpKbKUbEM3MzMxsLebLx2ZmZmbmSqGZmZmZTdBKoaTD02tZbpc0JmlmaV7HV7s0xTk2xblT0mJJf1Cad7Kk5ZKWSbpE0ro142wi6XJJ90i6W9If1omT5k+S9D+SrqqzX5K2l/TfqRzLJZ3UQ35y8ryrpJsl/VrSKU3zcvLcKU5OntvGSfOr5rllnBp57rRflfPcIu7Gkv5V0h2pHJXuGWkTa/90vi2X9OO6cVKsvSW9Jumomut3PE8y4tTObSlG1nfdJVal465CnMrnQpc4lc/NpvW+I+lJSctK0zaT9J+SfpZ+blozzllpv+6UtEjSJnXilOadIikkbV43jqQT03G0XNK8mvs1XdIteuP/tH0qxGl57OXmukOcrFx3Oxeq5rpTnNxc2wCMdweNbTpt3IA37nd8N3BPGp4E3A+8A5gC3AHs3iHOfsCmafhQYEka3hZ4EFgvjV8GHJcbJ40vAD6bhqcAm9SJk6b9BXAxcFWX/LTbr62BGWl4Q+DemvnJzfNbgb2BEjKAeQAAButJREFUvwZOKU3PzXPLODXy3DZOZp7b7VduntvFycpzi7hfAc5Mw1sAvwSm1DjfNgH+F9ihUd7cGE37dD3FTc5H1YzR8TzJKEft3Nb9rrvEqnTcVYhT+VzoECPr3Gxa9/3ADGBZado84LQ0fFrjuKwR50PA5DR8Zt04afr2FDfiPwxsXrM8BwD/BfxeGu96brSJcy1waBqeBdxQ99jLzXWHOFm57nQu5OS6Q3myc+1P/z8TsqUwIl6MdFQA6/NGh4tVXu1SjrM4Ip5No7dQ9NPTMBlYT9JkYCod+u9pF0fSRhS/AL6dlns1Ip6rUx5J2wF/AlzYbv1ucSJiRUQsTcMrgbspfvnnlic3z09GxK3Aqhazc/LcMk6NPLctT2aeW8apked25cnKc6vQwIaSRPGH1C+B32Ss33AMcEVEPNIob40YDScCPwJqx+hy3lbVa24bZcn6rtvJOe66xMk6F7qofG6WRcSNFMda2eEUlVXSzyPqxImIayOicQxX+u7blAfgHOBLVOiwt0Oc44G/jYhfp2W6Htdt4gSwURremAq57nDsZeW6XZzcXHc5FyrnukOc7Fxb/03ISiGApCMl3QNcDXw6Td4WeLS02GNU/wX9GeDfASLiceBs4BFgBUX/PdfmxqFohXgK+G66LHShpPVrxAE4l+Kk+m3F9dvFAUDSjsCewJIacXrJ8+t6zHNZL3luVjfPLdXIc1mveT4P2I3iP5i7gJMios5+7QJsKukGSbdJ+kSNGEjaFjgSuKDO+m20PL4r6MsxXNbjd92v464v50Ifz82GLSP1gZZ+vrWHWA2fpt53j6TZwOMRcUePZdgF+CNJSyT9WNLeNeN8AThL0qMUef9yzspNx17tXHc4hrNyXY7TS66bytOvXFsPJmylMCIWRcSuFH8FfSNNrvXaFkkHUPzncmoa35Tir623A9sA60v6WG4cir+0ZwDzI2JP4CWK5vzc8hwGPBkRt3Vbt0t5GtM3oGit+UJEvFAjTq08t4hbK88t1Mpzi/LUynOHeFl5bhWixbScPB8M3E6R2+nAeaklKddk4D0ULVkHA1+VtEuNOOcCp0bEazXW/R3tju+qq7eYVrv/rV6+6z4fd/06F/p1bg6EpNMpWr1/WGPdqcDpwF/2oSiTgU2BfYEvApellvlcxwMnR8T2wMmklt4q+vB7pmOc3FyX46T1auW6RXn6lWvrwYSpFEo6Id2Ee7ukbRrTU1P8Tunm1a6vbWmOI+ndFJdsDo+IZ9JiBwEPRsRTEbEKuILiPqbcOI8Bj0VE46+uyyl+YefGeR8wW9JDFJe5DpT0gxpxkLQOxYn2w4i4olueO+xXVp6bt5Nk57lNnOw8t4mTnec2cbLz3GG/sl5DVI4LnEBx2Tci4j6Ke8R27bR+mzhPAP8RES9FxNPAjUClhzua4uwFXJryexTwT5K6XkZsjtPp+M5Q6xVPbcrW8buuoOtxl6HruVBR13Mz0y8kbQ2Qfta+9Cfpk8BhwLERUacivxNFZfeOlPPtgKWStqoR6zHeOMd+StHS2/WhlRY+SZFjgH+muL2hqzbHXnau2x3DubluEadWrtuUp1+5tl7EBLixsfkDvJM3HjSZATxO8Zf/ZOABioOwcfP4Hh3i7EDRm/d+TdPfCyynuI9GFPdlnJgbJ837CfD7aXgEOKtOnNIy+9P9AYh2+yXge8C5FfPcLk5WnkvrjfDmBymy8twuTm6eO8XJyXOH/crKc4c4tfJcWn8+MJKGt0znSdcb6lvE2Q24LpVnKrAMeFdunKaYF1H/QZOu50mFGD3lttfvukO8ysddhxjZ50KLGLXOzdL6O/LmBynO4s0PP8yrGecQioeetsjcnzfFaZr3UNXzokV5Pgd8PQ3vQnFLgmrEuRvYPw1/ELit7rGXm+sOcbJyXeVcqJLrDuWplWt/+vsZ9wK0LFRxuWg5xaWxm4GZpXmzKJ5Wuh84vUucC4FnU5zbgbHSvK8B91D8B/h90hNPNeJMB8aAO4F/IT01mRuntMz+dK8UtowDzKS4RHZnad6smvuVk+etKP7KewF4Lg1vVCPPneLk5LltnMw8t4xTI8+d9qtynlvE3Ybiqca7Un4/1sM590WK/yCWUVzO6fUcvoj6lcKu50nFOLVzW4qR9V1XiNf1uKsQo/K50CVO5XOzab1LKO5DXJWO5c8A0yj+sPhZ+rlZzTj3UVQGGrm+oE6cpvkPUe3p41blmQL8IOVoKXBgzTgzgdso/jhZAryn7rGXm+sOcbJyXeVcqJLrDuXJzrU//f/4NXdmZmZmNnHuKTQzMzOz8eNKoZmZmZm5UmhmZmZmrhSamZmZGa4UmpmZmRmuFJqZmZkZrhSamZmZGfD/4F92/nrMTbQAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 842.4x595.44 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# get codon names (keys) for y-axis labels\n",
    "y_axis_labels = populate_codon_idx_dict().keys()\n",
    "\n",
    "# get position values for x-axis labels (centered at start codon = 0)\n",
    "x_axis_labels = [i for i in range(-30,28)]\n",
    "df = pd.DataFrame(tis0, columns=x_axis_labels, index=y_axis_labels)\n",
    "\n",
    "# plot heatmap\n",
    "a4_dims = (11.7, 8.27)\n",
    "fig, ax = plt.subplots(figsize=a4_dims)\n",
    "sns.heatmap(ax=ax, data=df, cmap=\"YlGnBu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TIS features extracted -- 279.2417 seconds\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>3703</th>\n",
       "      <th>3704</th>\n",
       "      <th>3705</th>\n",
       "      <th>3706</th>\n",
       "      <th>3707</th>\n",
       "      <th>3708</th>\n",
       "      <th>3709</th>\n",
       "      <th>3710</th>\n",
       "      <th>3711</th>\n",
       "      <th>3712</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AF_RS00005.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AF_RS00005.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AF_RS00005.2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AF_RS00005.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AF_RS00005.4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3713 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0     1     2     3     4     5     6     7     8     9     ...  \\\n",
       "0  AF_RS00005.0     0     0     0     0     0     0     0     0     0  ...   \n",
       "1  AF_RS00005.1     0     0     0     1     0     0     0     0     0  ...   \n",
       "2  AF_RS00005.2     1     0     0     0     0     0     0     0     0  ...   \n",
       "3  AF_RS00005.3     0     0     0     0     0     0     0     0     0  ...   \n",
       "4  AF_RS00005.4     0     0     0     0     0     0     0     0     0  ...   \n",
       "\n",
       "   3703  3704  3705  3706  3707  3708  3709  3710  3711  3712  \n",
       "0     0     0     0     0     0     0     0     0     0     0  \n",
       "1     0     0     0     0     0     0     0     0     0     0  \n",
       "2     0     0     0     0     0     0     0     0     0     0  \n",
       "3     0     0     0     0     0     0     0     0     0     0  \n",
       "4     0     0     0     0     0     0     0     0     0     0  \n",
       "\n",
       "[5 rows x 3713 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tis_temp = []\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "for record in _orfs:\n",
    "    name = record.name\n",
    "    strand = record.features[0].strand\n",
    "    start  = record.features[0].location.start.position\n",
    "\n",
    "    # create frame (58x64) for TIS pos\n",
    "    seq60 = (dna[0][start-30:start+30] if strand == 1 else dna[1][::-1][start-30:start+30][::-1]).seq\n",
    "\n",
    "    frame = [TRICODON_IDX_DICT[seq60[i:i+3]] for i in range(0,len(seq60)-2)]\n",
    "    tis = list(to_categorical(frame, num_classes=64).astype(int).T.reshape(-1))\n",
    "    tis_temp.append([name] + tis)\n",
    "    \n",
    "print('TIS features extracted -- %s seconds' % (round(time.time() - start_time, 4)))\n",
    "\n",
    "_tis = pd.DataFrame(tis_temp).reset_index(drop=True)\n",
    "_tis.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
